design/multi-checkout.md

```diff
@@ -0,0 +1,180 @@
+0. If there is exactly one `checkout: none`, then we'll preserve the existing behavior: no repositories are synced or checked out.
+0. If there is exactly one `checkout` step that isn't `none`, then we'll use existing single-checkout behavior but check out that repository rather than `self`.
+
+### More than one checkout step
+0. If there are multiple checkout steps and one or more are `checkout: none`, this is an error.
+It's unclear what the user meant for us to do, so we fail with a clear message.
+0. Otherwise, we are in multi-checkout mode.
+
+### Behavior changes in multi-checkout mode
+
+#### Checkout directory
+By default, all repositories will be checked out as if they've been `git clone`d into the `Pipeline.Workspace` directory.
+Git uses the last part of the path, minus any trailing slashes, spaces, and the string `.git`.
+(We have YAML syntax for selecting a different path, and if the user specifies that, we'll use their name.)
+This changes the default directory for the `self` repo: it would be `$(Pipeline.Workspace)/s` if it were the only `checkout` step.
+
+#### System variables
+`Build.SourcesDirectory` points to the directory where the `self` repo was checked out whether in single- or multi-checkout mode.
+`System.DefaultWorkingDirectory` points to the same place as `Pipeline.Workspace` in multi-checkout mode (which is different than single-checkout mode).
+We believe this preserves the intended semantics of these variables, keeping as many scripts and tasks working as possible.
+
+## Resource specification
+
+The `resources` block already allows a `repositories` list.
+Today, those repositories hold only YAML templates, so we can't blindly check them out automatically.
+That's why an explicit `checkout` step is also required.
+
+However, in multi-checkout scenarios, this leads to unnecessary verbosity.
+First you have to mention the repository up top, give it a name, and then you check it out by name.
+To combat this, we'll add an inline repository syntax.
+Inline syntax will cover the `type`, `name`, and `ref` of a repository resource.
+(If other changes are needed - such as `path` or `endpoint` - users must fall back to explicit resource syntax.)
+
+## Syntax examples
+
+### Basic syntax with inline resource
+
+The most basic, and probably most common form of multi-checkout:
+```yaml
+steps:
+- checkout: self
+- checkout: git://MyProject/MyToolsRepo
+```
+
+This grabs a tools repository from Azure Repos and checks out its default branch, typically `master`.
+Assuming this is in a repository called "MySourceRepo", the resulting directory structure on a hosted agent would be:
+
+```
+_w/
+   1/
+     MySourceRepo/
+         (files from MySourceRepo)
+     MyToolsRepo/
+         (files from MyToolsRepo)
+```
+
+### Basic syntax, GitHub source provider
+
+The same syntax works for GitHub:
+
+```yaml
+steps:
+- checkout: self
+- checkout: github://MyGitHubOrg/MyToolsRepo
```

Comment:
Note that here no endpoint is defined, so this must be a public GitHub repo.
We should also verify that it works with public GitHub repos.